### PR TITLE
HPCC-25212 Add support for unity builds in CMake

### PR DESCRIPTION
**Pull request by jpmcmu**
WIP to reduce compile time of the platform. At the moment working on testing impact of enabling unity builds.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [X] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [X] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [X] The commit message is properly formatted and free of typos.
  - [X] The commit message title makes sense in a changelog, by itself.
  - [X] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the CONTRIBUTORS document.
- [X] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jpmcmu** [comment]
Creating this to share some experiments to reduce compile time of the platform. Currently have enabled unity builds on some of the sub-projects. Enabled unity build on some of the projects that don't require as much work to fix duplicate symbols etc. Enabling on all projects would further cut down compile time another 4-5 minutes.

Precompiled headers for things like jlib would also probably have a decent impact on compile time.

Timing of make -j6 on my VM:

Without Unity Builds:
real    31m1.788s
user    115m24.239s
sys     6m0.177s

Unity Builds:
real    21m47.372s
user    80m47.214s
sys     4m0.005s

**Comment by jpmcmu** [comment]
@rpastrana please review

**Comment by jpmcmu** [comment]
@richardkchapman @ghalliday 

Ran some comparisons between ninja and make. Ninja didn't really provide any improvement on my VM. It could be that  Ninja would provide a benefit with more cores? (Disabled AWS, Azure, WS_SQL to get ninja to work, fixes to those projects wouldn't be too difficult if we want to get ninja working for them)

Make without Unity Build
real    20m1.892s
user    74m3.783s
sys     3m44.617s

Ninja without Unity Build
real    19m43.717s
user    74m49.448s
sys     3m34.738s

----------------------------

Ninja with Unity Build
real    15m39.071s
user    59m46.666s
sys     2m31.265s

Make with Unity Build
real    15m36.859s
user    57m37.442s
sys     2m33.697s

**Comment by ghalliday** [comment]
Thanks, James.  What is your system configuration?
Gavin

________________________________

**Comment by jpmcmu** [comment]
@ghalliday The VM I was compiling in has 4 cores and 8Gb

**Comment by ghalliday** [comment]
A similar set of timings from a machine with 16 cores, and plenty of memory:

Make without Unity Build
real	6m6.840s
user	70m35.321s
sys	4m29.613s

Ninja without Unity Build
real	4m52.003s
user	71m1.937s
sys	4m19.930s

With unity (with cppunit also disabled)

Make with Unity Build
real	5m47.557s
user	51m57.114s
sys	2m59.358s

Ninja with Unity Build
real	4m12.480s
user	52m40.994s
sys	2m55.314s


Which suggests if you have plenty of cores that ninja has a much bigger effect than unity!

**Comment by jpmcmu** [comment]
@ghalliday Ah interesting. Out of curiosity what was the -j option you gave make? I have heard that make works best with a j value of 150% of the core count. Curious to see if that still holds up at 16 cores

**Comment by ghalliday** [comment]
That was -j18.  There are a few stages where the cpu dropped significantly below all cores in use. ninja seemed to keep them all active.

**Comment by jpmcmu** [comment]
Hmm I wonder if the 150% -j24 would do a better job keeping all cores utilized?

**Comment by ghalliday** [comment]
I don't think so...  -j24:

real	6m0.963s
user	68m31.254s
sys	4m19.242s

make -j (overcommitted with no number)

real	6m30.859s
user	69m36.479s
sys	4m24.983s


I think the issue is that some of the projects can't use all the threads, and in those cases make doesn't do a great job of building other things in parallel.

**Comment by ghalliday** [comment]
My impression is that unity, ninja and possibly precompiled headers may all have a significant impact on build times.  I think for this PR I would add an option USE_UNITY into commonSetup which can then be used to easily toggle the functionality on and off.

**Comment by ghalliday** [comment]
@jpmcmu I hadn't assigned back to you.  I think it is worth pursing, and adding a global option to enable it - so that it is easy to turn on and off.  It will need rebasing once changes are upmerged from 7.12.x to master.

**Comment by ghalliday** [comment]
@jpmcmu are you planning on clening up this branch, or would you prefer me to do it?

**Comment by jpmcmu** [comment]
@ghalliday I will clean it up and get it ready to go this week; I wouldn't want to add more work to your list. There are still a few projects that don't have unity build support, but it probably makes more sense to add unity support for those projects that in a future PR.

**Comment by ghalliday** [comment]
yes, I would gradually add support to other parts.  This should be enough to speed some systems up siginficantly.

**Comment by jpmcmu** [comment]
@ghalliday please review

File: ecl/eclcmd/eclcmd_common.cpp [hunk 1]
```diff
@@ -31,6 +31,7 @@ static StringBuffer eclccpath;
 
 const char *queryEclccPath(bool optVerbose)
 {
+    static StringBuffer eclccpath;
```
Comment (ghalliday) [inline ecl/eclcmd/eclcmd_common.cpp:34]: Should the definition on line 30 have been deleted?

File: system/jlib/jatomic.hpp [hunk 2]
```diff
@@ -197,10 +197,13 @@ typedef volatile long atomic_t;
 
 #elif defined(__GNUC__)
 
-typedef struct { volatile int counter; } atomic_t;
 #define ATOMIC_INIT(i)          { (i) }
 #define atomic_read(v)          ((v)->counter)
+
```
Comment (ghalliday) [inline system/jlib/jatomic.hpp:202]: This change doesn't match the one in eclrtl_imp.hpp.  That added

#ifndef atomic_type_defined

All of this code should be deleted anyway, but I have created a separate jira for that.

**Review by ghalliday — CHANGES_REQUESTED** [review]
@jpmcmu It looks good.  The cmake docs suggest it should be implemented slightly differently though:

> Projects should not directly set the UNITY_BUILD property or its associated CMAKE_UNITY_BUILD variable to true. .... However, it IS recommended to set the UNITY_BUILD target property to false if it is known that enabling unity builds for the target can lead to problems.

https://cmake.org/cmake/help/latest/prop_tgt/UNITY_BUILD.html#prop_tgt:UNITY_BUILD

I think this would mean all the places where it is set to true are removed (and a different mechanism is used to disable it?)

**Comment by ghalliday** [comment]
@jpmcmu see Richard's PR https://github.com/hpcc-systems/HPCC-Platform/pull/15785

**Pull request closed without merging**